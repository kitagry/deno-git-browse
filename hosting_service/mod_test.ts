import { assertRejects } from "https://deno.land/std@0.202.0/assert/mod.ts";
import { assertSnapshot } from "https://deno.land/std@0.202.0/testing/snapshot.ts";
import { getHostingService, UnsupportedHostingServiceError } from "./mod.ts";

Deno.test("getHostingService", async (t) => {
  await t.step("throws error for unsupported hosting service", async () => {
    const url = new URL("https://example.com/lambdalisue/gin.vim");
    await assertRejects(
      () => {
        return getHostingService(url);
      },
      UnsupportedHostingServiceError,
    );
  });

  const urls = [
    new URL("ssh://git@github.com/lambdalisue/gin.vim"),
    new URL("https://github.com/lambdalisue/gin.vim"),
  ];
  for (const url of urls) {
    const svc = await getHostingService(url);

    await t.step(`getHomeURL for ${url}`, async () => {
      const result = svc.getHomeURL(url);
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getCommitURL for ${url}`, async () => {
      const result = svc.getCommitURL(url, "main");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getTreeURL for ${url}`, async () => {
      const result = svc.getTreeURL(url, "main", "denops/gin");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getBlobURL for ${url}`, async () => {
      const result = svc.getBlobURL(url, "main", "denops/gin/main.ts");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getBlobURL with lineStart for ${url}`, async () => {
      const result = svc.getBlobURL(url, "main", "denops/gin/main.ts", {
        range: 10,
      });
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(
      `getBlobURL with lineStart/lineEnd for ${url}`,
      async () => {
        const result = svc.getBlobURL(url, "main", "denops/gin/main.ts", {
          range: [10, 20],
        });
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      },
    );

    if (svc.getPullRequestURL) {
      await t.step(`getPullRequestURL for ${url}`, async () => {
        const result = svc.getPullRequestURL!(url, 1);
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      });
    }
  }
});

Deno.test("getHostingService with alias", async (t) => {
  const urls = [
    new URL("https://my-github.com/lambdalisue/gin.vim"),
  ];
  const aliases = {
    "my-github.com": "github.com",
  };

  for (const url of urls) {
    const svc = await getHostingService(url, { aliases });

    await t.step(`getHomeURL for ${url}`, async () => {
      const result = svc.getHomeURL(url);
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getCommitURL for ${url}`, async () => {
      const result = svc.getCommitURL(url, "main");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getTreeURL for ${url}`, async () => {
      const result = svc.getTreeURL(url, "main", "denops/gin");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getBlobURL for ${url}`, async () => {
      const result = svc.getBlobURL(url, "main", "denops/gin/main.ts");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getBlobURL with lineStart for ${url}`, async () => {
      const result = svc.getBlobURL(url, "main", "denops/gin/main.ts", {
        range: 10,
      });
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(
      `getBlobURL with lineStart/lineEnd for ${url}`,
      async () => {
        const result = svc.getBlobURL(url, "main", "denops/gin/main.ts", {
          range: [10, 20],
        });
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      },
    );

    if (svc.getPullRequestURL) {
      await t.step(`getPullRequestURL for ${url}`, async () => {
        const result = svc.getPullRequestURL!(url, 1);
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      });
    }
  }
});
