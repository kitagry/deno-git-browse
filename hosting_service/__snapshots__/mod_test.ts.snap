export const snapshot = {};

snapshot[`getHostingService 1`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 2`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/commit/2120c87633f4059656e8aada8eebe62768892b46",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 3`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/tree/v0.1.0/bin",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 4`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 5`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 6`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10-L20",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 7`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/pull/1",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 8`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 9`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/commit/2120c87633f4059656e8aada8eebe62768892b46",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 10`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/tree/v0.1.0/bin",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 11`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 12`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 13`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10-L20",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 14`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/pull/1",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 1`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 2`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/commit/2120c87633f4059656e8aada8eebe62768892b46",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 3`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/tree/v0.1.0/bin",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 4`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 5`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 6`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10-L20",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 7`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/pull/1",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;
